### PR TITLE
[WNMGDS-532] Add CMSDS core link for child design systems

### PR DESCRIPTION
**Pull request by line47**
### Added
- Added CMSDS documentation flag to include a link to the core CMSDS 
- Added maturity status badge to all component pages

### Changed
- Updated doc site header to include link to CMSDS for component and pattern pages - this link shows up on the child design system only 

### Screenshot
Screen Shot 2020-12-04 at 5 02 19 PM

### How to test
- setup a child design system in the package.json to this branch with `file:../design-system/package/design-system/`
- run `git clean -fdx` and then `yarn install` Install the local DS branch in the child ds
- Visit component and pattern pages 
- See that pages that are from the core will have the CMSDS link 
- See that pages that are unique to the child design system will not have the CMSDS link

### CODE REVIEW

File: guides/WRITING-DOCUMENTATION.md [hunk 1]
```diff
@@ -77,6 +77,7 @@ To extend the default functionality of KSS, we've implemented support for custom
 | **`@responsive`**           | `@responsive`                                                | Renders breakpoint toggles for the markup example.                                                                                                     |
 | **`@status [NAME]`**        | `@status Draft`                                              | Displays a status badge. `[NAME]` must be one of the following values: `Draft`, `Work in progress`, `Ready`, `Deprecated`.                             |
 | **`@uswds [URL]`**          | `@uswds https://designsystem.digital.gov/components/button/` | `[URL]` is the link to the corresponding US Web Design System component.                                                                               |
+| **`@cmsds [URL]`**          | `@cmsds https://design.cms.gov/components/button/`           | `[URL]` is the link to the corresponding CMS Design System component.                                                                                  |
```
Comment (bernardwang) [inline guides/WRITING-DOCUMENTATION.md:80]: If we are automatically populating this for child design systems, do we still need to add this functionality?

File: packages/design-system-docs/src/pages/components/Alert/_Alert.docs.scss [hunk 2]
```diff
@@ -1,6 +1,8 @@
 /*
 Alert
 
+@status Ready
```
Comment (bernardwang) [inline packages/design-system-docs/src/pages/components/Alert/_Alert.docs.scss:4]: Is this just for testing?

Comment (bernardwang) [inline packages/design-system-docs/src/pages/components/Alert/_Alert.docs.scss:4]: Is there a particular reason why the Ready status is being added to each component? I know that each one is not in Draft, but I was under the impression it would be good to use for newly Ready components, and not necessarily every single component

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 3]
```diff
@@ -205,14 +207,28 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
+The inheritance of the USWDS and CMSDS link breaks becasuse the child version of the page takes precendence. 
+*/
+
+  // Get pages that come from core CMS Design System
+  const corePages = markdownSections.concat(kssSections).map((page) => {
+    if (
+      page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/components') ||
+      page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/patterns')
+    ) {
+      page.cmsds = `https://design.cms.gov/${page.referenceURI}`;
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:222]: this looks great!

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 4]
```diff
@@ -205,14 +207,28 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
   );
 
+  // 1. Are we only showing CMSDS link for patterns and components?
+
+  /* 2. If you create a duplicate page like Button at the child design system level and edit the first section of the documentation page. 
+The inheritance of the USWDS and CMSDS link breaks becasuse the child version of the page takes precendence. 
+*/
+
+  // Get pages that come from core CMS Design System
+  const corePages = markdownSections.concat(kssSections).map((page) => {
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:217]: `corePages` isnt really a great name for this variable, as it contains all the pages, just with some modifications for certain pages. I would rename it, or consider grouping it in with the `const pages` line below

Comment (bernardwang) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:217]: This is what I would suggest, it involves changing the names of 2 other variables from `pages` to `uniquePageSections` and from `nestedPages` to `pages`
```
const pageSections = markdownSections.concat(kssSections)
   .map(section => addCmsdsLink(section))

const uniquePageSections = uniquePages(pageSections).filter((page) => !page.hideSection);

const pages = await addTopLevelPages(pages).then(nestSections);
```

File: packages/design-system-docs/src/pages/components/Alert/_Alert.docs.scss (hunk 2 above)
Comment (line47) [reply packages/design-system-docs/src/pages/components/Alert/_Alert.docs.scss]: Yeah good point I'm removing them

File: guides/WRITING-DOCUMENTATION.md (hunk 1 above)
Comment (line47) [reply guides/WRITING-DOCUMENTATION.md:80]: No i don't think so

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js (hunk 4 above)
Comment (bernardwang) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:217]: On second thought maybe these names are better:
`allPageSections`
`uniquePageSections`
`nestedPageSections`

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx [hunk 5]
```diff
@@ -36,26 +54,27 @@ class PageHeader extends React.PureComponent {
...
+            id={this.props.reference}
+          />
           {this.statusBadge()}
-
+        </div>
+        <div className="ds-u-font-size--small">
           {this.guidanceLink()}
-
           {this.uswdsLink()}
+          {this.cmsdsLink()}
```
Comment (bernardwang) [inline packages/design-system-docs/src/scripts/components/PageHeader.jsx:69]: Could we put the CMSDS link first? We probably need to redo the logic here as well. Instead of trying to chain the two links together with two `span` elements, I would just write logic for two different cases: CMSDS + USWDS, and USWDS only

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 6]
```diff
@@ -26,6 +26,7 @@ const uniquePages = require('./uniquePages');
 const { get } = require('lodash');
 const { getDocsDirs } = require('../../common/getDirsToProcess');
 const { logTask } = require('../../common/logUtil');
+// const { elementTextContains } = require('selenium-webdriver/lib/until');
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:29]: I think this should be deleted

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 7]
```diff
@@ -180,6 +181,10 @@ async function generateExamplePages(pageSection, docsPath, sourceDir, options, c
 module.exports = async function generatePages(sourceDir, docsDir, options, changedPath) {
   logTask('📝 ', 'Generating documentation pages');
 
+   // Create HTML files for example and doc pages
+  const docsPath = path.join(docsDir, 'dist');
+  
+  // This gets doc site direcotires
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:187]: Can we remove this comment

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 8]
```diff
@@ -180,6 +181,10 @@ async function generateExamplePages(pageSection, docsPath, sourceDir, options, c
 module.exports = async function generatePages(sourceDir, docsDir, options, changedPath) {
   logTask('📝 ', 'Generating documentation pages');
 
+   // Create HTML files for example and doc pages
+  const docsPath = path.join(docsDir, 'dist');
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:185]: I'm not sure if this comment is accurate, `docsPath` is just the location of all the build files including HTML, CSS and JS files. I would personally move this variable declaration right before [this line](https://github.com/CMSgov/design-system/pull/890/files#diff-9af9efe78a2334efeb563a0d95bd8b68a4f0187123d4ad7d875ecde2ad3403d1R235), as its not actually used until `generateExamplePages` and `generateDocPages`

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 9]
```diff
@@ -201,28 +206,48 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
     )
   );
 
-  // Merge both sets of page sections into a single array.
+
+  const addCmsdsLink = function (page) {
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:210]: I dont think this syntax would work, as its combining a named function declaration with a variable declaration. You can take a look at the other functions in the same file as examples (i.e. `function changedFilter()`). These are two examples of how a function can be defined:

```
const addCmsdsLink = () => {};
function addCmsdsLink() {}
```

Comment (bernardwang) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:210]: I would also move this function code outside of the `generateDocPages` function. It can still be defined in the file, but it's cleaner if you put it elsewhere.

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 10]
```diff
@@ -201,28 +206,48 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
+    if (
+      page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/components') ||
+      page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/patterns')
+    ) {
+      page.cmsds = `https://design.cms.gov/${page.referenceURI}`;
+    }
+    return page;
+  };
+  // Get pages that come from core CMS Design System
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:219]: This comment isnt quite accurate either, I would remove it or clarify what this line is doing.

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 11]
```diff
@@ -201,28 +206,48 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
+  const pageSections = markdownSections.concat(kssSections).map((section) => addCmsdsLink(section));
+
+  // Merge both sets of KssSection objects into a single array of page parts.
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:222]: At this point the KssSection objects have already been merged (that happened with `markdownSections.concat(kssSections)`). I would remove this comment

File: packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js [hunk 12]
```diff
@@ -92,6 +92,11 @@ function processFlags(section) {
             // KSS converts the URL to an <a> element, so we grab just the URL
             section.uswds = hrefUrl(value);
             break;
+          case 'cmsds':
+            // CMS Design System URL
+            // KSS converts the URL to an <a> element, so we grab just the URL
+            section.cmsds = hrefUrl(value);
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js:98]: I dont think this is needed anymore right?

File: packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js [hunk 13]
```diff
@@ -11,8 +11,9 @@ function uniquePages(pages) {
   pages.forEach((page) => {
     if (routes[page.reference]) {
       if (routes[page.reference].source.path.match(/node_modules/)) {
-        // We override pages that come from `node_modules`
+        // We override existing pages that come from `node_modules` with pages from child design system
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js:14]: // We override pages that come from `node_modules` with pages from the child design system

Is more accurate. This `if` statement for overriding pages that already are in the `routes` map, but I'm not sure if thats important. Adding the `existing` word could be confusing

File: packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js [hunk 14]
```diff
@@ -11,8 +11,9 @@ function uniquePages(pages) {
...
         logTask('🖊  ', `Overriding ${page.reference || 'index'} page with ${page.source.path}`);
+        page.cmsds = routes[page.reference].cmsds;
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js:16]: I would add a comment here on why this line is necessary

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx (hunk 5 above)
Comment (line47) [reply packages/design-system-docs/src/scripts/components/PageHeader.jsx:66]: Thanks Bernard - I believe I have this fixed up now

**Comment by line47** [comment]
@bernardwang and @nichia this is ready for another review 👀 
Thank you!

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx [hunk 15]
```diff
@@ -3,21 +3,48 @@ import React from 'react';
 import classNames from 'classnames';
 
 class PageHeader extends React.PureComponent {
+  cmsdsAndUswdsLink() {
+    if (this.props.uswds && this.props.cmsds) {
+      return (
+        <span>
+          <span>View related guidance in the </span>
```
Comment (bernardwang) [inline packages/design-system-docs/src/scripts/components/PageHeader.jsx:10]: I dont think all these spans are required, you should be able to go with this:
```
<span> View related guidance in the <a href={this.props.cmsds}>CMS Design System</a> and <a href={this.props.uswds}>U.S. Web Design System</a>
</span>
```

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx [hunk 16]
```diff
@@ -36,26 +63,28 @@ class PageHeader extends React.PureComponent {
...
+            dangerouslySetInnerHTML={{ __html: this.props.header }}
+            id={this.props.reference}
+          />
           {this.statusBadge()}
-
+        </div>
+        <div className="ds-u-font-size--small">
           {this.guidanceLink()}
-
+          {this.cmsdsLink()}
```
Comment (bernardwang) [inline packages/design-system-docs/src/scripts/components/PageHeader.jsx:77]: Oh I completely forgot about the case where its possible to have a CMSDS link but not a USWDS, good catch. If thats the case, instead of having 3 functions for what's essentially 1 if else statement, can we organize this to be `{this.designSystemLinks()}`, where its defined as:
```
designSystemLinks () {
   const cmsdsLink = this.props.cmsds && (
       <a href={this.props.cmsds}>CMS Design System</a>
   );
   const uswdsLink = this.props.uswds && (
       <a href={this.props.uswds}>US Web Design System</a>
   );

   if (cmsds && uswdsLink) {
      return (<span> View related guidance in the {cmsdsLink} and {uswdsLink}</span>)
   } else if (cmsds) {
      return (<span> View related guidance in the {cmsdsLink}</span>)
   } else if (uswds) {
      return (<span> View related guidance in the {uswdsLink}</span>)
   } else {
      return
   }
}
```

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 17]
```diff
@@ -200,32 +200,51 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
       processKssSection(kssSection, options)
     )
   );
+  // Get page sections that come from core and child design systems
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:203]: `pageSections` is just all the doc site sections from KSS and markdown files, which may or may not be from the core or child DS. 

I think a better description here would be:
"Combine KSS and Markdown page sections and add cmsds links to for child design systems"

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 18]
```diff
@@ -200,32 +200,51 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
+
+  // Create HTML files for doc pages
+  const docPages = await generateDocPages(nestedPageSections, docsPath, options, changedPath);
   if (changedPath && docPages > 0) {
     logTask('📝 ', `Doc page updated from ${changedPath}`);
   } else if (!changedPath) {
     logTask('📝  ' + docPages, `Doc pages added to ${docsDir}`);
   }
 };
+function addCmsdsLink(page) {
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:242]: Can we move this higher up in the file? The `generatePages()` function is the main export, so it should be last. I would put this after `changedFilter`

Comment (bernardwang) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:242]: I think this function would benefit from a comment. Something like:

```
// Add a `cmsds` property to component and pattern page sections originating from the `@cmsgov/design-system-docs` NPM package.
// This automatically populates a link to the CMSDS for child design system components and patterns
```

File: packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js [hunk 19]
```diff
@@ -92,6 +92,11 @@ function processFlags(section) {
             // KSS converts the URL to an <a> element, so we grab just the URL
             section.uswds = hrefUrl(value);
             break;
+          // case 'cmsds':
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js:95]: Can we remove these?

File: packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js [hunk 20]
```diff
@@ -11,8 +11,10 @@ function uniquePages(pages) {
   pages.forEach((page) => {
     if (routes[page.reference]) {
       if (routes[page.reference].source.path.match(/node_modules/)) {
-        // We override pages that come from `node_modules`
+        // We override pages that come from node_modules with pages from the child design system
         logTask('🖊  ', `Overriding ${page.reference || 'index'} page with ${page.source.path}`);
+        // Creates cmsds property used on child design systems doc pages to show a link to the corresponding component from the core CMS Design System
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/uniquePages.js:16]: The `addCmsdsLink()` function in `generatePages` is what actually creates the property, this line is just to make sure that property is preserved when overriding. I would reword to be:

```
// Preserves the `cmsds` property created by `addCmsdsLink()` when overriding
// Even when a page is overridden by a child design system, we want to show the CMSDS link 
```

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 21]
```diff
@@ -200,32 +200,51 @@ module.exports = async function generatePages(sourceDir, docsDir, options, chang
...
+  if (
+    page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/components') ||
+    page.source.path.includes('node_modules/@cmsgov/design-system-docs/src/pages/patterns')
+  ) {
```
Comment (nichia) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:246]: Should we skip the markdown files with file extension `.md`? Can we change the search for core files to components and patterns with file extensions  `.docs.scss` ?
` const regexp = /^node_modules\/.*\/((components|patterns)\/.*\.docs\.scss)$/i;`
`  if (page.source.path.match(regexp)) {`

File: packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js (hunk 19 above)
Comment (line47) [reply packages/design-system-scripts/gulp/docs/generatePages/processKssSection.js:95]: Yeah - I missed those the first time, thanks!

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js (hunk 21 above)
Comment (line47) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:246]: I don't think we want to skip the markdown files here becasue this pattern page would then not be found 
`node_modules/@cmsgov/design-system-docs/src/pages/patterns/documentation-page.md` 

Am i thinking about this correctly?

Comment (nichia) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js]: You're right, do we want this pattern page to show the link?

Comment (line47) [reply packages/design-system-scripts/gulp/docs/generatePages/generatePages.js]: Yeah we do want the link to show on the documentation-page

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx [hunk 22]
```diff
@@ -33,29 +25,49 @@ class PageHeader extends React.PureComponent {
...
   }
 
+  designSystemLinks() {
+    const cmsdsLink = this.props.cmsds && <a href={this.props.cmsds}>CMS Design System</a>;
+    const uswdsLink = this.props.uswds && <a href={this.props.uswds}>U.S. Web Design System</a>;
+
+    if (this.props.cmsds && this.props.uswds) {
+      return (
+        <span>
+          {' '}
```
Comment (bernardwang) [inline packages/design-system-docs/src/scripts/components/PageHeader.jsx:35]: What's this space for?

File: packages/design-system-scripts/gulp/docs/generatePages/generatePages.js [hunk 23]
```diff
@@ -116,6 +116,18 @@ function changedFilter(page, changedPath) {
   return true;
 }
 
+// Add a `cmsds` property to component and pattern page sections originating from the `@cmsgov/design-system-docs` NPM package.
```
Comment (bernardwang) [inline packages/design-system-scripts/gulp/docs/generatePages/generatePages.js:119]: Thank you!

**Review by bernardwang — APPROVED** [review]
LGTM! great work on this scott, thanks for being patient with all the feedback!

**Review by nichia — APPROVED** [review]
Looks great!

File: packages/design-system-docs/src/scripts/components/PageHeader.jsx (hunk 22 above)
Comment (line47) [reply packages/design-system-docs/src/scripts/components/PageHeader.jsx:35]: It's not needed, thanks!
